File: mobile/ios/Air/SubModules/WalletCore/Helpers/TransferHelpers.swift
//
//  TransferHelpers.swift
//  MyTonWalletAir
//
//  Created by Sina on 12/20/24.
//

import WalletContext


public class TransferHelpers {
    private init() {}
    
    public struct ExplainedTransferFee {
        /** Whether the result implies paying the fee with a diesel */
        public let isGasless: Bool;
        /**
         * The fee that will be sent with the transfer. The wallet must have it on the balance to send the transfer.
         * Show this in the transfer form when the input amount is ≤ the balance, but the remaining balance can't cover the
         * full fee; show `realFee` otherwise. Undefined means that it's unknown.
         */
        public let fullFee: MFee?
        /**
         * The real fee (the full fee minus the excess). Undefined means that it's unknown. There is no need to fall back to
         * `fullFee` when `realFee` is undefined (because it's undefined too in this case).
         */
        public let realFee: MFee?
        /**
         * Whether the full token balance can be transferred despite the fee.
         * If yes, the fee will be taken from the transferred amount.
         */
        public let canTransferFullBalance: Bool
    };

    public static func shouldUseDiesel(input: MTransactionDraft) -> Bool {
        return input.diesel != nil && input.diesel?.status != .notAvailable
    }

    public static func explainApiTransferFee(chain: String, isNativeToken: Bool, input: MTransactionDraft) -> ExplainedTransferFee {
        return shouldUseDiesel(input: input) ? explainGaslessTransferFee(diesel: input.diesel!) : explainGasfullTransferFee(chain: chain,
                                                                                                                            isNativeToken: isNativeToken,
                                                                                                                            input: input)
    }
    
    // TODO: replace with normal impl without ?? 
    private static func explainGaslessTransferFee(diesel: MDiesel) -> ExplainedTransferFee {
        let isStarsDiesel = diesel.status == .starsFee
        let realFeeInDiesel = convertFee(amount: diesel.realFee, exampleFromAmount: diesel.nativeAmount, exampleToAmount: (isStarsDiesel ? diesel.starsAmount : diesel.tokenAmount) ?? 0);
        // Cover as much displayed real fee as possible with diesel, because in the excess it will return as the native token.
        let dieselRealFee = min(diesel.tokenAmount ?? 0, realFeeInDiesel);
        // Cover the remaining real fee with the native token.
        let nativeRealFee = convertFee(amount: realFeeInDiesel - dieselRealFee,
                                       exampleFromAmount: diesel.tokenAmount ?? 0,
                                       exampleToAmount: diesel.nativeAmount);

        return ExplainedTransferFee(
            isGasless: true,
            fullFee: MFee(
                precision: .lessThan,
                terms: .init(
                    token: diesel.tokenAmount,
                    native: diesel.remainingFee,
                    stars: diesel.starsAmount
                ),
                nativeSum: diesel.realFee
            ),
            realFee: MFee(
                precision: .approximate,
                terms: .init(token: isStarsDiesel ? nil : dieselRealFee,
                             native: nativeRealFee,
                             stars: isStarsDiesel ? dieselRealFee : nil),
                nativeSum: diesel.realFee
            ),
            canTransferFullBalance: false
        )
    }
    
    private static func explainGasfullTransferFee(chain: String, isNativeToken: Bool, input: MTransactionDraft) -> ExplainedTransferFee {
        let fullFee = MFee(precision: input.realFee == input.fee ? .exact : .lessThan,
                           terms: .init(token: nil,
                                        native: input.fee,
                                        stars: nil),
                           nativeSum: input.fee ?? 0)

        return ExplainedTransferFee(isGasless: false,
                                    fullFee: input.fee != nil ? fullFee : nil,
                                    realFee: input.realFee != nil ? MFee(precision: input.realFee == input.fee ? .exact : .approximate,
                                                                         terms: .init(token: nil,
                                                                                      native: input.realFee,
                                                                                      stars: nil),
                                                                         nativeSum: input.fee ?? 0) : fullFee,
                                    canTransferFullBalance: chain == "ton" && isNativeToken)
    }
    
    private static func convertFee(amount: BigInt,
                                   exampleFromAmount: BigInt,
                                   exampleToAmount: BigInt) -> BigInt {
        if exampleFromAmount == 0 {
            return 0
        }
        return amount * exampleToAmount / exampleFromAmount
    }

    public static func getMaxTransferAmount(tokenBalance: BigInt?,
                                            isNativeToken: Bool,
                                            fullFee: MFee.FeeTerms?,
                                            canTransferFullBalance: Bool) -> BigInt? {
        guard let tokenBalance, tokenBalance > 0 else {
            return nil
        }

        // Returning the full balance when the fee is unknown for a better UX
        if canTransferFullBalance || fullFee == nil {
            return tokenBalance
        }
        
        var fee = fullFee?.token ?? 0
        if isNativeToken {
            fee += fullFee?.native ?? 0
        }
        return max(tokenBalance - fee, 0)
    }
}

File: mobile/android/air/SubModules/UISend/src/main/java/org/mytonwallet/app_air/uisend/send/helpers/TransferHelpers.kt
package org.mytonwallet.app_air.uisend.send.helpers

import org.mytonwallet.app_air.walletcore.models.ExplainedTransferFee
import org.mytonwallet.app_air.walletcore.models.MFee
import org.mytonwallet.app_air.walletcore.moshi.IDiesel
import org.mytonwallet.app_air.walletcore.moshi.MApiCheckTransactionDraftResult
import org.mytonwallet.app_air.walletcore.moshi.MDieselStatus
import java.math.BigInteger

class TransferHelpers private constructor() {

    companion object {
        private fun shouldUseDiesel(input: MApiCheckTransactionDraftResult): Boolean {
            return input.diesel != null && input.diesel?.status != MDieselStatus.NOT_AVAILABLE && input.diesel?.tokenAmount != null
        }

        fun explainApiTransferFee(
            chain: String,
            isNativeToken: Boolean,
            input: MApiCheckTransactionDraftResult
        ): ExplainedTransferFee {
            return if (shouldUseDiesel(input)) {
                explainGaslessTransferFee(input.diesel!!)
            } else {
                explainGasfullTransferFee(chain, isNativeToken, input)
            }
        }

        private fun explainGaslessTransferFee(diesel: IDiesel): ExplainedTransferFee {
            val isStarsDiesel = diesel.status == MDieselStatus.STARS_FEE
            val dieselAmount = if (isStarsDiesel) diesel.starsAmount!! else diesel.tokenAmount!!
            val realFeeInToken =
                convertFee(
                    diesel.realFee ?: BigInteger.ZERO,
                    diesel.nativeAmount!!,
                    dieselAmount
                )
            val tokenRealFee = minOf(dieselAmount, realFeeInToken)
            val nativeRealFee =
                convertFee(
                    realFeeInToken - tokenRealFee,
                    dieselAmount,
                    diesel.nativeAmount ?: BigInteger.ZERO
                )

            return ExplainedTransferFee(
                isGasless = true,
                fullFee = MFee(
                    precision = MFee.FeePrecision.LESS_THAN,
                    terms = MFee.FeeTerms(
                        token = diesel.tokenAmount,
                        native = diesel.remainingFee,
                        stars = diesel.starsAmount
                    ),
                    nativeSum = (diesel.nativeAmount ?: BigInteger.ZERO) +
                        (diesel.remainingFee ?: BigInteger.ZERO),
                ),
                realFee = MFee(
                    precision = MFee.FeePrecision.APPROXIMATE,
                    terms = MFee.FeeTerms(
                        token = if (isStarsDiesel) null else tokenRealFee,
                        native = nativeRealFee,
                        stars = if (isStarsDiesel) tokenRealFee else null
                    ),
                    nativeSum = diesel.realFee
                ),
                canTransferFullBalance = false,
                excessFee = (diesel.nativeAmount ?: BigInteger.ZERO) +
                    (diesel.remainingFee ?: BigInteger.ZERO) -
                    (diesel.realFee ?: BigInteger.ZERO)
            )
        }

        private fun explainGasfullTransferFee(
            chain: String,
            isNativeToken: Boolean,
            input: MApiCheckTransactionDraftResult
        ): ExplainedTransferFee {
            val fullFee = MFee(
                precision = if (input.realFee == input.fee) MFee.FeePrecision.EXACT else MFee.FeePrecision.LESS_THAN,
                terms = MFee.FeeTerms(
                    token = null,
                    native = input.fee,
                    stars = null
                ),
                nativeSum = input.fee
            )

            return ExplainedTransferFee(
                isGasless = false,
                fullFee = input.fee?.let { fullFee },
                realFee = input.realFee?.let {
                    MFee(
                        precision = if (input.realFee == input.fee) MFee.FeePrecision.EXACT else MFee.FeePrecision.APPROXIMATE,
                        terms = MFee.FeeTerms(
                            token = null,
                            native = input.realFee,
                            stars = null,
                        ),
                        nativeSum = input.realFee
                    )
                } ?: fullFee,
                canTransferFullBalance = chain == "ton" && isNativeToken,
                excessFee = (input.fee ?: BigInteger.ZERO) - (input.realFee ?: BigInteger.ZERO)
            )
        }

        private fun convertFee(
            amount: BigInteger,
            exampleFromAmount: BigInteger,
            exampleToAmount: BigInteger
        ): BigInteger {
            return amount * exampleToAmount / exampleFromAmount
        }

        fun getMaxTransferAmount(
            tokenBalance: BigInteger?,
            isNativeToken: Boolean,
            fullFee: MFee.FeeTerms?,
            canTransferFullBalance: Boolean
        ): BigInteger? {
            if (tokenBalance == null || tokenBalance <= BigInteger.ZERO) {
                return null
            }

            if (canTransferFullBalance || fullFee == null) {
                return null
            }

            var fee = fullFee.token ?: BigInteger.ZERO
            if (isNativeToken) {
                fee += fullFee.native ?: BigInteger.ZERO
            }
            return maxOf(tokenBalance - fee, BigInteger.ZERO)
        }
    }
}
